core/src/swarm/registry.rs

```diff
@@ -178,4 +206,23 @@ mod tests {
         }
         QuickCheck::new().quickcheck(property as fn(Vec<u32>) -> bool)
     }
+
+    #[test]
+    fn old_reports_disappear() {
+        let mut addresses = Addresses::default();
+
+        // Add an address a single time.
+        let single: Multiaddr = "/tcp/2108".parse().unwrap();
+        addresses.add(single.clone());
+        assert!(addresses.iter().find(|a| **a == single).is_some());
+
+        // Then fill `addresses` with random stuff.
+        let other: Multiaddr = "/tcp/120".parse().unwrap();
+        for _ in 0 .. 2000 {
+            addresses.add(other.clone());
+        }
+
+        // Check that `single` disappeared from the list.
+        assert!(addresses.iter().find(|a| **a == single).is_none());
+    }
```

Comment:
Since the important property is that each score reflects the number of reports in the last `n` reports, I would replace this test with a corresponding property test:

```rust
    #[test]
    fn record_score_equals_last_n_reports() {
        use multiaddr::Protocol;
        use quickcheck::{Arbitrary, Gen};
        use rand::Rng;
        use std::num::NonZeroUsize;

        #[derive(PartialEq, Eq, Clone, Hash, Debug)]
        struct Ma(Multiaddr);

        impl Arbitrary for Ma {
            fn arbitrary<G: Gen>(g: &mut G) -> Self {
                Ma(Protocol::Tcp(g.gen::<u16>() % 16).into())
            }
        }

        fn property(xs: Vec<Ma>, n: u8) -> bool {
            let n = std::cmp::max(n, 1);
            let mut addresses = Addresses::new(NonZeroUsize::new(30).unwrap(), usize::from(n));
            for Ma(a) in &xs {
                addresses.add(a.clone())
            }
            for r in &addresses.registry {
                let count = xs.iter()
                    .rev()
                    .take(usize::from(n))
                    .filter(|Ma(x)| x == &r.addr)
                    .count();
                if r.score as usize != count {
                    return false
                }
            }
            true
        }

        QuickCheck::new().quickcheck(property as fn(Vec<Ma>, u8) -> bool)
    }
```